base/compiler/abstractinterpretation.jl

```diff
@@ -372,37 +372,54 @@ function precise_container_type(@nospecialize(arg), @nospecialize(typ), vtypes::
     elseif tti0 <: Array
         return Any[Vararg{eltype(tti0)}]
     else
-        return Any[abstract_iteration(typ, vtypes, sv)]
+        return abstract_iteration(typ, vtypes, sv)
     end
 end
 
 # simulate iteration protocol on container type up to fixpoint
 function abstract_iteration(@nospecialize(itertype), vtypes::VarTable, sv::InferenceState)
     tm = _topmod(sv)
```

Comment:
As implemented right now, this actually always uses `Base`, not `_topmod(sv)` to find `iterate`. We should probably redefine this as `if isdefined(Main, :Base) || !isdefined(Main.Base, :iterate) || !isconst(Main.Base, :iterate)` to be pedantic and precise about correctness here

Reply:
Ok, will adjust then.

Reply:
Um, are you sure? If I look at e.g. `Core.Compiler.append_any(1)`, it seems to be calling `Core.Compiler.iterate`. And that's what we want to reflect, no?

Reply:
I don’t think there’s any callers of `Core.Compiler.append_any`. The implementation of _apply right now always uses Base